### PR TITLE
Map::Entry methods to recover key and value together

### PR DESCRIPTION
**Pull request by seanmonstar**
See https://github.com/rust-lang/rust/issues/32281#issuecomment-213066344

### CODE REVIEW

**Comment by sfackler** [comment]
Compilation errors: https://travis-ci.org/rust-lang/rust/builds/126956602

**Comment by seanmonstar** [comment]
Fixed

**Comment by seanmonstar** [comment]
An alternative that may be nice (and keep the method names consistent) is to `impl Into<K> for VacantEntry` and `impl Into<(K, V)> for OccupiedEntry`.

**Comment by alexcrichton** [comment]
The libs team discussed this during triage the other day and the conclusion was that this seems like fine functionality to have but we likely want to closely consider the naming.

I would personally propose `into_key` for `VacantEntry` and `remove_pair` for `OccupiedEntry` as the former conveys the transfer of ownership and the latter mirrors the existing `remove` function  along with the returned pair (although the term "pair" is kinda unfortunate).

**Comment by alexcrichton** [comment]
cc @rust-lang/libs

**Comment by arielb1** [comment]
Maybe Python-style `remove_item`?

**Comment by seanmonstar** [comment]
I've changed the names to `into_key` and `remove_pair`.

File: src/libcollections/btree/map.rs [hunk 1]
```diff
@@ -1898,6 +1898,12 @@ impl<'a, K: Ord, V> VacantEntry<'a, K, V> {
         &self.key
     }
 
+    /// Take ownership of the key.
+    #[unstable(feature = "map_entry_keys", issue = "32281")]
```
Comment (alexcrichton) [inline src/libcollections/btree/map.rs]: Could these point to a new issue? Pointing to a close issue  may unfortunately mean that these never come up for stabilization :(

Comment (seanmonstar) [inline src/libcollections/btree/map.rs]: Oh right. Probably need a new feature as well, cause `map_entry_keys` is now stable. So just create a new issue "Tracking for map_entry_recover_keys" or some such?

Comment (alexcrichton) [inline src/libcollections/btree/map.rs]: Yeah I can be sure to tag it appropriately and such.

**Comment by seanmonstar** [comment]
Added `map_entry_recover_keys` feature, pointing at https://github.com/rust-lang/rust/issues/34285

**Pull request merged**